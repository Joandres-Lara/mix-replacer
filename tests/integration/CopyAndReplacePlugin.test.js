const mix = require("laravel-mix");
const fs = require("fs");
const path = require("path");
const CopyAndReplacePlugin = require("../../src/CopyAndReplacePlugin");
const {getTempDir, withHashQuery} = require("../helpers");

const public_urls = [
	"/images/touch/homescreen48.png",
	"/images/touch/homescreen72.png",
	"/images/touch/homescreen96.png",
	"/images/touch/homescreen168.png",
	"/images/touch/homescreen192.png",
	"/images/mstile-150x150.png",
	"/images/mstile-70x70.png",
	"/images/mstile-310x310.png",
	"/js/jquery.js",
	"/js/my-background.js",
	"/images/button/geo-38.png",
	"/popup/geo.html",
];

describe("Normal flow", () => {
	test.each(["with versioning", "without versioning"])(
		"%s",
		(param, done) => {
			const with_versioning = param === "with versioning";
			const tmpdir = getTempDir();
			const plugin = new CopyAndReplacePlugin();
			const public_dir = `${tmpdir}${SEP}public`;
			const Stats = {compilation: {assets: {}}};

			// Setting up a mock manifest object
			Mix.manifest.manifest = public_urls.reduce((accumulator, public_url) => {

				accumulator[public_url] = with_versioning
				                          ? withHashQuery(public_url)
				                          : public_url;

				return accumulator;
			}, {});

			const callback = () => {

				try {
					expect(Stats).toStrictEqual({
						compilation: {
							assets: {
								"/browserconfig.xml": {size: expect.any(Function), emitted: true},
								"/config.json": {size: expect.any(Function), emitted: true},
								"/config_files/manifest.json": {size: expect.any(Function), emitted: true},
							},
						},
					});

					const assets = Stats.compilation.assets;

					const assets_data = Object.keys(assets).reduce(
						(accumulator, public_url) => {
							const target_public_path = path.join(public_dir, public_url);

							accumulator[public_url] = {
								is_expected_size: assets[public_url].size() === fs.statSync(target_public_path).size,
								content: fs.readFileSync(target_public_path).toString(),
							};

							return accumulator;
						}, {});

					expect({
						assets_data: assets_data,
						manifest: require(`${public_dir}${SEP}mix-manifest.json`),
					}).toMatchSnapshot();

					done();
				} catch (e) {
					done(e);
				} finally {
					Mix.components.has.mockRestore();
				}

			};

			// This mock is necessary since mix.version() is not available for us right now
			jest.spyOn(Mix.components, "has").mockReturnValue(with_versioning);

			mix.setPublicPath(public_dir);
			plugin.register("tests/integration/resources/*", public_dir);
			plugin.register("tests/integration/resources/sub1/**/*", `${public_dir}${SEP}config_files`);
			plugin.onCompilerDone(Stats, callback);
		},
	);
});

test("Throws Error", () => {
	const plugin = new CopyAndReplacePlugin();

	expect(() => {
		const tmpdir = getTempDir();
		plugin.register("non/existant/**/*", tmpdir);
	}).toThrow(/didn't yield any results/g);
});
