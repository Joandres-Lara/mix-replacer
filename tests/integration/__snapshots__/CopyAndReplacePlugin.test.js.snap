// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normal flow with versioning 1`] = `
Object {
  "assets_data": Object {
    "/browserconfig.xml": Object {
      "content": "<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<browserconfig>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png?id=5db8be17609750169201\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#70ba84</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#471e7d</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png?id=40a5f274c19553d0a7a0\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png?id=7b5572e4d23d5f53bf92\\"/>
            <TileColor>#ffffff</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#ffffff</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#e3bf86</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png?id=5db8be17609750169201\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"{{
            /images/mstile-310x310.png
            }}\\"/>
            <TileColor>#ffffff</TileColor>
        </tile>
    </msapplication>
</browserconfig>
",
      "is_expected_size": true,
    },
    "/config.json": Object {
      "content": "{
	\\"name\\": \\"HackerWeb\\",
	\\"short_name\\": \\"HackerWeb\\",
	\\"start_url\\": \\".\\",
	\\"display\\": \\"standalone\\",
	\\"background_color\\": \\"#fff\\",
	\\"description\\": \\"A simply readable Hacker News app.\\",
	\\"icons\\": [
		{
			\\"src\\": \\"/images/touch/homescreen48.png?id=7980c3918f3b58e0bc7a\\",
			\\"sizes\\": \\"48x48\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen72.png?id=c6c0ba94beb35d51b9b2\\",
			\\"sizes\\": \\"72x72\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen96.png?id=382f98b5fe07525e9941\\",
			\\"sizes\\": \\"96x96\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen144.png\\",
			\\"sizes\\": \\"144x144\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen168.png?id=4065147e3a975df29dd2\\",
			\\"sizes\\": \\"168x168\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen192.png?id=afed5bee88d35c60877d\\",
			\\"sizes\\": \\"192x192\\",
			\\"type\\": \\"image/png\\"
		}
	],
	\\"related_applications\\": [
		{
			\\"platform\\": \\"play\\",
			\\"url\\": \\"https://play.google.com/store/apps/details?id=cheeaun.hackerweb\\"
		}
	]
}
",
      "is_expected_size": true,
    },
    "/config_files/manifest.json": Object {
      "content": "{
	\\"browser_specific_settings\\": {
		\\"gecko\\": {
			\\"id\\": \\"addon@example.com\\",
			\\"strict_min_version\\": \\"42.0\\"
		}
	},
	\\"background\\": {
		\\"scripts\\": [
			\\"/js/jquery.js?id=6db478e854f4572c8444\\",
			\\"/js/my-background.js?id=43879dd6558e5b09b302\\"
		]
	},
	\\"browser_action\\": {
		\\"default_icon\\": {
			\\"19\\": \\"button/geo-19.png\\",
			\\"38\\": \\"/images/button/geo-38.png?id=b0227ed3cc1850898d31\\"
		},
		\\"default_title\\": \\"Whereami?\\",
		\\"default_popup\\": \\"popup/geo.html\\"
	},
	\\"commands\\": {
		\\"toggle-feature\\": {
			\\"suggested_key\\": {
				\\"default\\": \\"Ctrl+Shift+Y\\",
				\\"linux\\": \\"Ctrl+Shift+U\\"
			},
			\\"description\\": \\"Send a 'toggle-feature' event\\"
		}
	},
	\\"content_security_policy\\": \\"script-src 'self' https://example.com; object-src 'self'\\",
	\\"content_scripts\\": [
		{
			\\"exclude_matches\\": [
				\\"*://developer.mozilla.org/*\\"
			],
			\\"matches\\": [
				\\"*://*.mozilla.org/*\\"
			],
			\\"js\\": [
				\\"/js/borderify.js\\"
			]
		}
	],
	\\"default_locale\\": \\"en\\",
	\\"description\\": \\"...\\",
	\\"icons\\": {
		\\"48\\": \\"icon.png\\",
		\\"96\\": \\"icon@2x.png\\"
	},
	\\"manifest_version\\": 2,
	\\"name\\": \\"...\\",
	\\"page_action\\": {
		\\"default_icon\\": {
			\\"19\\": \\"button/geo-19.png\\",
			\\"38\\": \\"/images/button/geo-38.png?id=b0227ed3cc1850898d31\\"
		},
		\\"default_title\\": \\"Whereami?\\",
		\\"default_popup\\": \\"/popup/geo.html\\"
	},
	\\"permissions\\": [
		\\"webNavigation\\"
	],
	\\"version\\": \\"0.1\\",
	\\"user_scripts\\": {
		\\"api_script\\": \\"apiscript.js\\"
	},
	\\"web_accessible_resources\\": [
		\\"{{/images
		/my-image.png}}\\"
	]
}
",
      "is_expected_size": true,
    },
  },
  "manifest": Object {
    "/browserconfig.xml": "/browserconfig.xml?id=5026c6b745cdabc97f53",
    "/config.json": "/config.json?id=3c264d5f7e068cd665f9",
    "/config_files/manifest.json": "/config_files/manifest.json?id=924bdab720275b6c0aad",
    "/images/button/geo-38.png": "/images/button/geo-38.png?id=b0227ed3cc1850898d31",
    "/images/mstile-150x150.png": "/images/mstile-150x150.png?id=5db8be17609750169201",
    "/images/mstile-310x310.png": "/images/mstile-310x310.png?id=7b5572e4d23d5f53bf92",
    "/images/mstile-70x70.png": "/images/mstile-70x70.png?id=40a5f274c19553d0a7a0",
    "/images/touch/homescreen168.png": "/images/touch/homescreen168.png?id=4065147e3a975df29dd2",
    "/images/touch/homescreen192.png": "/images/touch/homescreen192.png?id=afed5bee88d35c60877d",
    "/images/touch/homescreen48.png": "/images/touch/homescreen48.png?id=7980c3918f3b58e0bc7a",
    "/images/touch/homescreen72.png": "/images/touch/homescreen72.png?id=c6c0ba94beb35d51b9b2",
    "/images/touch/homescreen96.png": "/images/touch/homescreen96.png?id=382f98b5fe07525e9941",
    "/js/jquery.js": "/js/jquery.js?id=6db478e854f4572c8444",
    "/js/my-background.js": "/js/my-background.js?id=43879dd6558e5b09b302",
    "/popup/geo.html": "/popup/geo.html?id=d7bc85eb914b5a219de1",
  },
}
`;

exports[`Normal flow without versioning 1`] = `
Object {
  "assets_data": Object {
    "/browserconfig.xml": Object {
      "content": "<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<browserconfig>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#70ba84</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#471e7d</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#ffffff</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#ffffff</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"/images/mstile-310x310.png\\"/>
            <TileColor>#e3bf86</TileColor>
        </tile>
    </msapplication>
    <msapplication>
        <tile>
            <square70x70logo src=\\"/images/mstile-70x70.png\\"/>
            <square150x150logo src=\\"/images/mstile-150x150.png\\"/>
            <wide310x150logo src=\\"/images/mstile-310x150.png\\"/>
            <square310x310logo src=\\"{{
            /images/mstile-310x310.png
            }}\\"/>
            <TileColor>#ffffff</TileColor>
        </tile>
    </msapplication>
</browserconfig>
",
      "is_expected_size": true,
    },
    "/config.json": Object {
      "content": "{
	\\"name\\": \\"HackerWeb\\",
	\\"short_name\\": \\"HackerWeb\\",
	\\"start_url\\": \\".\\",
	\\"display\\": \\"standalone\\",
	\\"background_color\\": \\"#fff\\",
	\\"description\\": \\"A simply readable Hacker News app.\\",
	\\"icons\\": [
		{
			\\"src\\": \\"/images/touch/homescreen48.png\\",
			\\"sizes\\": \\"48x48\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen72.png\\",
			\\"sizes\\": \\"72x72\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen96.png\\",
			\\"sizes\\": \\"96x96\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen144.png\\",
			\\"sizes\\": \\"144x144\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen168.png\\",
			\\"sizes\\": \\"168x168\\",
			\\"type\\": \\"image/png\\"
		},
		{
			\\"src\\": \\"/images/touch/homescreen192.png\\",
			\\"sizes\\": \\"192x192\\",
			\\"type\\": \\"image/png\\"
		}
	],
	\\"related_applications\\": [
		{
			\\"platform\\": \\"play\\",
			\\"url\\": \\"https://play.google.com/store/apps/details?id=cheeaun.hackerweb\\"
		}
	]
}
",
      "is_expected_size": true,
    },
    "/config_files/manifest.json": Object {
      "content": "{
	\\"browser_specific_settings\\": {
		\\"gecko\\": {
			\\"id\\": \\"addon@example.com\\",
			\\"strict_min_version\\": \\"42.0\\"
		}
	},
	\\"background\\": {
		\\"scripts\\": [
			\\"/js/jquery.js\\",
			\\"/js/my-background.js\\"
		]
	},
	\\"browser_action\\": {
		\\"default_icon\\": {
			\\"19\\": \\"button/geo-19.png\\",
			\\"38\\": \\"/images/button/geo-38.png\\"
		},
		\\"default_title\\": \\"Whereami?\\",
		\\"default_popup\\": \\"popup/geo.html\\"
	},
	\\"commands\\": {
		\\"toggle-feature\\": {
			\\"suggested_key\\": {
				\\"default\\": \\"Ctrl+Shift+Y\\",
				\\"linux\\": \\"Ctrl+Shift+U\\"
			},
			\\"description\\": \\"Send a 'toggle-feature' event\\"
		}
	},
	\\"content_security_policy\\": \\"script-src 'self' https://example.com; object-src 'self'\\",
	\\"content_scripts\\": [
		{
			\\"exclude_matches\\": [
				\\"*://developer.mozilla.org/*\\"
			],
			\\"matches\\": [
				\\"*://*.mozilla.org/*\\"
			],
			\\"js\\": [
				\\"/js/borderify.js\\"
			]
		}
	],
	\\"default_locale\\": \\"en\\",
	\\"description\\": \\"...\\",
	\\"icons\\": {
		\\"48\\": \\"icon.png\\",
		\\"96\\": \\"icon@2x.png\\"
	},
	\\"manifest_version\\": 2,
	\\"name\\": \\"...\\",
	\\"page_action\\": {
		\\"default_icon\\": {
			\\"19\\": \\"button/geo-19.png\\",
			\\"38\\": \\"/images/button/geo-38.png\\"
		},
		\\"default_title\\": \\"Whereami?\\",
		\\"default_popup\\": \\"/popup/geo.html\\"
	},
	\\"permissions\\": [
		\\"webNavigation\\"
	],
	\\"version\\": \\"0.1\\",
	\\"user_scripts\\": {
		\\"api_script\\": \\"apiscript.js\\"
	},
	\\"web_accessible_resources\\": [
		\\"{{/images
		/my-image.png}}\\"
	]
}
",
      "is_expected_size": true,
    },
  },
  "manifest": Object {
    "/browserconfig.xml": "/browserconfig.xml",
    "/config.json": "/config.json",
    "/config_files/manifest.json": "/config_files/manifest.json",
    "/images/button/geo-38.png": "/images/button/geo-38.png",
    "/images/mstile-150x150.png": "/images/mstile-150x150.png",
    "/images/mstile-310x310.png": "/images/mstile-310x310.png",
    "/images/mstile-70x70.png": "/images/mstile-70x70.png",
    "/images/touch/homescreen168.png": "/images/touch/homescreen168.png",
    "/images/touch/homescreen192.png": "/images/touch/homescreen192.png",
    "/images/touch/homescreen48.png": "/images/touch/homescreen48.png",
    "/images/touch/homescreen72.png": "/images/touch/homescreen72.png",
    "/images/touch/homescreen96.png": "/images/touch/homescreen96.png",
    "/js/jquery.js": "/js/jquery.js",
    "/js/my-background.js": "/js/my-background.js",
    "/popup/geo.html": "/popup/geo.html",
  },
}
`;
